File: tests/benchmarks/optimized_loop_benchmark.lua
#!/usr/bin/env lua
local TRIALS = 5
local ITER_SIMPLE = 5000000
local NEST_OUTER = 800
local NEST_INNER = 800
local ARRAY_LENGTH = 2048
local ARRAY_REPEATS = 512

local seed_values = {}
for i = 0, ARRAY_LENGTH - 1 do
    seed_values[#seed_values + 1] = i * 3
end

local total_simple = 0.0
local total_nested = 0.0
local total_array = 0.0
local checksum = 0

print("=== Lua Optimized Loop Benchmark ===")
print("trials:", TRIALS)

for trial = 0, TRIALS - 1 do
    local start_simple = os.clock()
    local simple_sum = 0
    for i = 0, ITER_SIMPLE - 1 do
        simple_sum = simple_sum + i
    end
    local elapsed_simple = os.clock() - start_simple
    total_simple = total_simple + elapsed_simple

    local start_nested = os.clock()
    local nested_acc = 0
    for outer = 0, NEST_OUTER - 1 do
        local inner = 0
        while inner < NEST_INNER do
            local combined = outer * inner
            if combined % 2 == 0 then
                nested_acc = nested_acc + combined
            else
                nested_acc = nested_acc - inner
            end
            inner = inner + 1
        end
    end
    local elapsed_nested = os.clock() - start_nested
    total_nested = total_nested + elapsed_nested

    local start_array = os.clock()
    local array_total = 0
    local repeat_count = 0
    while repeat_count < ARRAY_REPEATS do
        for idx = 1, #seed_values do
            array_total = array_total + seed_values[idx]
        end
        repeat_count = repeat_count + 1
    end
    local elapsed_array = os.clock() - start_array
    total_array = total_array + elapsed_array

    checksum = checksum + simple_sum + nested_acc + array_total

    print("trial", trial, "simple:", elapsed_simple, "nested:", elapsed_nested, "array:", elapsed_array)
end

local trials_f = TRIALS + 0.0
print("average_simple:", total_simple / trials_f)
print("average_nested:", total_nested / trials_f)
print("average_array:", total_array / trials_f)
print("checksum:", checksum)
print("=== Benchmark complete ===")

File: tests/benchmarks/optimized_loop_benchmark.js
#!/usr/bin/env node
const { performance } = require('perf_hooks');

const TRIALS = 5;
const ITER_SIMPLE = 5_000_000;
const NEST_OUTER = 800;
const NEST_INNER = 800;
const ARRAY_LENGTH = 2048;
const ARRAY_REPEATS = 512;

const seedValues = new Int32Array(ARRAY_LENGTH);
for (let i = 0; i < ARRAY_LENGTH; i++) {
  seedValues[i] = i * 3;
}

function elapsedSeconds(start) {
  return Number(performance.now() - start) / 1000;
}

let totalSimple = 0;
let totalNested = 0;
let totalArray = 0;
let checksum = 0;

console.log("=== JavaScript Optimized Loop Benchmark ===");
console.log("trials:", TRIALS);

for (let trial = 0; trial < TRIALS; trial++) {
  const startSimple = performance.now();
  let simpleSum = 0;
  for (let i = 0; i < ITER_SIMPLE; i++) {
    simpleSum += i;
  }
  const elapsedSimple = elapsedSeconds(startSimple);
  totalSimple += elapsedSimple;

  const startNested = performance.now();
  let nestedAcc = 0;
  for (let outer = 0; outer < NEST_OUTER; outer++) {
    let inner = 0;
    while (inner < NEST_INNER) {
      const combined = outer * inner;
      if ((combined & 1) === 0) {
        nestedAcc += combined;
      } else {
        nestedAcc -= inner;
      }
      inner += 1;
    }
  }
  const elapsedNested = elapsedSeconds(startNested);
  totalNested += elapsedNested;

  const startArray = performance.now();
  let arrayTotal = 0;
  let repeat = 0;
  while (repeat < ARRAY_REPEATS) {
    for (let idx = 0; idx < seedValues.length; idx++) {
      arrayTotal += seedValues[idx];
    }
    repeat += 1;
  }
  const elapsedArray = elapsedSeconds(startArray);
  totalArray += elapsedArray;

  checksum += simpleSum + nestedAcc + arrayTotal;

  console.log(
    "trial",
    trial,
    "simple:",
    elapsedSimple,
    "nested:",
    elapsedNested,
    "array:",
    elapsedArray
  );
}

const trialsF = TRIALS;
console.log("average_simple:", totalSimple / trialsF);
console.log("average_nested:", totalNested / trialsF);
console.log("average_array:", totalArray / trialsF);
console.log("checksum:", checksum);
console.log("=== Benchmark complete ===");
